File: medicine-search-backend/src/main/java/com/medicalsystem/medicine_search/service/InventoryService.java
package com.medicalsystem.medicine_search.service;

import com.medicalsystem.medicine_search.dto.InventoryRequestDTO;
import com.medicalsystem.medicine_search.dto.InventoryResponseDTO;
import com.medicalsystem.medicine_search.entity.Inventory;
import com.medicalsystem.medicine_search.entity.Medicine;
import com.medicalsystem.medicine_search.entity.Pharmacy;

import java.util.List;

public interface InventoryService {

    List<InventoryResponseDTO> getByMedicine(Medicine medicineId);

    List<InventoryResponseDTO> getByPharmacy(Pharmacy pharmacyId);

    List<InventoryResponseDTO> searchByPharmacyName(String pharmacyName);

//    InventoryResponseDTO getByMedicineAndPharmacy(Medicine medicineId, Pharmacy pharmacyId);

    List<InventoryResponseDTO> searchByMedicineNameSorted(String medicineName);

    List<InventoryResponseDTO> getAllInventories();

    InventoryResponseDTO getInventoryById(Long id);

    List<InventoryResponseDTO> searchByMedicineName(String medicineName);

    List<InventoryResponseDTO> searchByLocation(String location);

    List<InventoryResponseDTO> getAvailableStock();

//    InventoryResponseDTO saveInventory(Inventory inventory);

    InventoryResponseDTO updateInventory(Long id, InventoryRequestDTO inventoryRequestDTO);

    void deleteInventory(Long id);
}

File: medicine-search-frontend/src/app/core/services/inventory/inventory.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

export interface PharmacyDTO {
  pharmacyId: number;
  pharmacyName: string;
  pharmacyLocation: string;
}

export interface MedicineDTO {
  medicineId: number;
  medicineName: string;
}

export interface InventoryResponseDTO {
  id: number;
  pharmacy: PharmacyDTO;
  medicine: MedicineDTO;
  quantity: number;
  price: number;
  lastUpdated?: string;
}

export interface InventoryRequestDTO {
  pharmacyName: string;
  pharmacyLocation: string;
}

@Injectable({
  providedIn: 'root',
})
export class InventoryService {
  private baseUrl = `${environment.apiUrl}/inventory`; // change to deployed backend URL

  constructor(private http: HttpClient) {}

  getAllInventories(): Observable<InventoryResponseDTO[]> {
    return this.http.get<InventoryResponseDTO[]>(this.baseUrl);
  }

  searchByPharmacy(pharmacyName: string): Observable<InventoryResponseDTO[]> {
    return this.http.get<InventoryResponseDTO[]>(`${this.baseUrl}/search/pharmacy?pharmacyName=${pharmacyName}`);
  }

  searchByMedicine(medicineName: string): Observable<InventoryResponseDTO[]> {
    return this.http.get<InventoryResponseDTO[]>(`${this.baseUrl}/search/medicine?medicineName=${medicineName}`);
  }

  searchByLocation(location: string): Observable<InventoryResponseDTO[]> {
    return this.http.get<InventoryResponseDTO[]>(
      `${this.baseUrl}/search/location?location=${location}`,
    );
  }

  getAvailable(): Observable<InventoryResponseDTO[]> {
    return this.http.get<InventoryResponseDTO[]>(`${this.baseUrl}/available`);
  }

  createInventory(data: any): Observable<InventoryResponseDTO> {
    return this.http.post<InventoryResponseDTO>(this.baseUrl, data);
  }

  updateInventory(id: number, data: any): Observable<InventoryResponseDTO> {
    return this.http.put<InventoryResponseDTO>(`${this.baseUrl}/${id}`, data);
  }

  deleteInventory(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  refreshInventories(): Observable<InventoryResponseDTO[]> {
    return this.getAllInventories()
  }
}
